### PR TITLE
MEPTS-409 MOH Cohort Report - F Indicator

### PR DESCRIPTION
**Pull request by simaomazive**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java [hunk 1]
```diff
@@ -83,8 +90,23 @@ public CohortDefinition getE() {
...
+    sqlCohortDefinition.setQuery(
+        ResumoTrimestralQueries.getPatientsWhoReceivedOneViralLoadResult(
+            hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId(),
+            hivMetadata.getHivViralLoadConcept().getConceptId(),
+            hivMetadata.getBeyondDetectableLimitConcept().getConceptId()));
+
+    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+    cd.addSearch("E", mapStraightThrough(cohortE));
+    cd.addSearch("FI", mapStraightThrough(sqlCohortDefinition));
+    cd.setCompositionString("E AND NOT FI");
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:109]: The specs says to filter not to remove F1

Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:109]: Please call this E _cohortE_ and sqlCohortDefinition call it _viralLoadResult_

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java [hunk 2]
```diff
@@ -83,8 +90,23 @@ public CohortDefinition getE() {
...
+    sqlCohortDefinition.setName(
+        "Patients in the 1st line treatment who received one Viral load result");
+    sqlCohortDefinition.addParameter(new Parameter("onOrAfter", "onOrAfter", Date.class));
+    sqlCohortDefinition.addParameter(new Parameter("onOrBefore", "onOrBefore", Date.class));
+    sqlCohortDefinition.addParameter(new Parameter("location", "location", Location.class));
+    sqlCohortDefinition.setQuery(
+        ResumoTrimestralQueries.getPatientsWhoReceivedOneViralLoadResult(
+            hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId(),
+            hivMetadata.getHivViralLoadConcept().getConceptId(),
+            hivMetadata.getBeyondDetectableLimitConcept().getConceptId()));
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:104]: @simaomazive please separate this into a SqlCohortDefinition function

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 3]
```diff
@@ -3,4 +3,19 @@
...
+  public static String getPatientsWhoReceivedOneViralLoadResult(
+      int encounterType, int viralLoad, int viralLoadQualitative) {
+    String query =
+        "SELECT p.patient_id FROM "
+            + "patient p JOIN encounter e "
+            + "ON p.patient_id = e.patient_id JOIN obs o "
+            + "ON o.encounter_id = e.encounter_id "
+            + "WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND "
+            + "e.encounter_type = %d AND (o.concept_id = %d AND o.value_numeric IS NOT NULL) "
+            + "OR (o.concept_id = %d AND o.value_numeric IS NOT NULL) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:17]: We no longer using this type of passing the parameters, please adjust accordingly, use 

> StringSubstitutor

Comment (simaomazive) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java (hunk 2 above)
Comment (simaomazive) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:104]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java (hunk 1 above)
Comment (simaomazive) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java]: Done

Comment (simaomazive) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java]: Done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java [hunk 4]
```diff
@@ -83,13 +83,39 @@ public CohortDefinition getE() {
     return cd;
   }
 
+  /**
+   * SqlCohortDefinition for Viral Load Result
+   *
+   * @return
+   */
+  public CohortDefinition getViralLoadResult() {
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:91]: @simaomazive please use a more suggestive name like 
```suggestion
  public CohortDefinition getNumberOfPatientsInFichaClinicaWithViralLoadResults() {
```

Comment (psmatsinhe) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralCohortQueries.java:91]: @simaomazive like here https://github.com/jembi/openmrs-module-eptsreports/blob/develop/api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java#L101

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 5]
```diff
@@ -8,6 +8,26 @@
...
 
+  public static String getPatientsWhoReceivedOneViralLoadResult(
+      int encounterType, int viralLoad, int viralLoadQualitative) {
+    String query =
+        "SELECT p.patient_id FROM "
+            + "patient p JOIN encounter e "
+            + "ON p.patient_id = e.patient_id JOIN obs o "
+            + "ON o.encounter_id = e.encounter_id "
+            + "WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND "
+            + "e.encounter_type = ${encounterType} AND (o.concept_id = ${viralLoad} AND o.value_numeric IS NOT NULL) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:19]: @simaomazive the reason we using this stringsubstitutor is to be able to ready the query without having to go and see what encounterType he is talking about...
Please change this name to be more suggestive like 
```suggestion
            + "e.encounter_type = ${adultoSeguimentoEncounterType} AND (o.concept_id = ${viralLoadConcept} AND o.value_numeric IS NOT NULL) "
```

Comment (psmatsinhe) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:19]: See here https://github.com/jembi/openmrs-module-eptsreports/blob/develop/api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoMensalCohortQueries.java#L916

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 6]
```diff
@@ -8,6 +8,26 @@
...
+            + "OR (o.concept_id = ${viralLoadQualitative} AND o.value_numeric IS NOT NULL) "
+            + "AND e.encounter_datetime <= :onOrBefore group by patient_id";
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("encounterType", encounterType);
+    valuesMap.put("viralLoad", viralLoad);
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:25]: Here
```suggestion
    valuesMap.put("viralLoadConcept", viralLoadConcept);
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 7]
```diff
@@ -8,6 +8,26 @@
...
+  public static String getPatientsWhoReceivedOneViralLoadResult(
+      int encounterType, int viralLoad, int viralLoadQualitative) {
+    String query =
+        "SELECT p.patient_id FROM "
+            + "patient p JOIN encounter e "
+            + "ON p.patient_id = e.patient_id JOIN obs o "
+            + "ON o.encounter_id = e.encounter_id "
+            + "WHERE p.voided = 0 AND e.voided = 0 AND o.voided = 0 AND "
+            + "e.encounter_type = ${encounterType} AND (o.concept_id = ${viralLoad} AND o.value_numeric IS NOT NULL) "
+            + "OR (o.concept_id = ${viralLoadQualitative} AND o.value_numeric IS NOT NULL) "
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:20]: Here 
```suggestion
            + "OR (o.concept_id = ${viralLoadQualitativeConcept} AND o.value_numeric IS NOT NULL) "
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 8]
```diff
@@ -8,6 +8,26 @@
...
+            + "AND e.encounter_datetime <= :onOrBefore group by patient_id";
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("encounterType", encounterType);
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:24]: Here 
```suggestion
    valuesMap.put("adultoSeguimentoEncounterType", adultoSeguimentoEncounterType);
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 9]
```diff
@@ -8,6 +8,26 @@
...
+    valuesMap.put("viralLoad", viralLoad);
+    valuesMap.put("viralLoadQualitative", viralLoadQualitative);
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:26]:
```suggestion
    valuesMap.put("viralLoadQualitativeConcept", viralLoadQualitativeConcept);
```

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java [hunk 10]
```diff
@@ -8,6 +8,26 @@
 @Component
 public class ResumoTrimestralQueries {
 
+  public static String getPatientsWhoReceivedOneViralLoadResult(
+      int encounterType, int viralLoad, int viralLoadQualitative) {
```
Comment (psmatsinhe) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/ResumoTrimestralQueries.java:12]:
```suggestion
      int adultoSeguimentoEncounterType, int viralLoadConcept, int viralLoadQualitativeConcept) {
```

**Pull request merged**